nemo/export/trt_llm/nemo/nemo.py

```diff
@@ -0,0 +1,283 @@
+        else:
+            tokenizer.add_special_tokens({"pad_token": "<pad>"})
+    else:
+        tokenizer.add_special_tokens({"pad_token": "<pad>"})
+
+    if not hasattr(tokenizer, "bos_token"):
+        if hasattr(tokenizer.tokenizer, "bos_id") and tokenizer.tokenizer.bos_id() > 0:
+            tokenizer.bos_token = tokenizer.tokenizer.id_to_piece(tokenizer.tokenizer.bos_id())
+        else:
+            tokenizer.add_special_tokens({"bos_token": "<bos>"})
+    else:
+        tokenizer.add_special_tokens({"bos_token": "<s>"})
+
+    if not hasattr(tokenizer, "eos_token"):
+        if hasattr(tokenizer.tokenizer, "eos_id") and tokenizer.tokenizer.eos_id() > 0:
+            tokenizer.eos_token = tokenizer.tokenizer.id_to_piece(tokenizer.tokenizer.eos_id())
+        else:
+            tokenizer.add_special_tokens({"eos_token": "<eos>"})
+    else:
+        tokenizer.add_special_tokens({"eos_token": "</s>"})
+
+
+# TODO: remove tar.extractall usage before releasing with KitMaker
+def unpack_nemo_ckpt(
+    nemo_archive_path: typing.Union[str, pathlib.Path], out_dir_path: typing.Union[str, pathlib.Path],
+):
+    nemo_archive_path = pathlib.Path(nemo_archive_path)
+    if not nemo_archive_path.exists():
+        raise FileNotFoundError(f"{nemo_archive_path} does not exist")
+
+    for tar_mode in ["r:", "r:gz"]:
+        try:
+            with tarfile.open(nemo_archive_path, mode=tar_mode) as tar_file:
+
+                def is_within_directory(directory, target):
+                    abs_directory = os.path.abspath(directory)
+                    abs_target = os.path.abspath(target)
+
+                    prefix = os.path.commonprefix([abs_directory, abs_target])
+
+                    return prefix == abs_directory
+
+                def safe_members(tar_file):
+                    members = []
+                    for member in tar_file.getmembers():
+                        member_path = os.path.join(out_dir_path, member.name)
+                        if not is_within_directory(out_dir_path, member_path):
+                            raise Exception("Attempted Path Traversal in Tar File")
+                        members.append(member)
+                    return members
+
+                tar_file.extractall(
+                    out_dir_path, members=safe_members(tar_file), numeric_owner=False
+                )  # nosec - tar path has been validated.
+
+            return out_dir_path
+        except tarfile.ReadError:
+            pass
+
+    raise RuntimeError(f"Could not unpack {nemo_archive_path}")
+
+
+def extract_layers_with_prefix(model_, prefix):
+    length_to_trim = len(prefix)
+    model_state = model_.get("state_dict", model_)
+    return {key[length_to_trim:]: model_state[key] for key in model_state.keys() if prefix in key}
+
+
+class UnpackedNemoCheckpointDir:
+    def __init__(
+        self, checkpoints_dir: typing.Union[str, pathlib.Path], load_checkpoints_to_cpu: bool = False,
```

Comment:
Please see if you can depend on SaveRestoreConnector or subclasses to deal with package handling of nemo files. The directory structure is not guaranteed to remain the same, so its safer to access via that system.

Reply:
Same reason above.